[quant][pt2e][be] Remove _input_output_share_observers and _reuse_input_obs_or_fq from QuantizationAnnotation

jerryzh168 opened the pull request:
Summary:
att, after we support SharedQuantizationSpec we don't need these things anymore, this PR refactors the
uses of _input_output_share_observers to SharedQuantizationSpec

Test Plan:
```
buck2 test mode/opt caffe2/test:quantization_pt2e -- 'caffe2/test:quantization_pt2e'
```

Reviewed By: andrewor14

Differential Revision: D46301342

leslie-fang-intel commented on torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py:
```diff
@@ -571,17 +572,20 @@ def _annotate_maxpool2d(
-            input_act_qspec = get_input_act_qspec(quantization_config)
-            output_act_qspec = get_output_act_qspec(quantization_config)
+            # only annotate maxpool when the input node is annotated
+            if "quantization_annotation" not in input_act.meta or not input_act.meta["quantization_annotation"]._annotated:
```
To to more specific, I think we should check the `input_act.meta["quantization_annotation"].output_qspec is not None`.  Since we may only annotate the input but not output of the `input_act`.

jerryzh168 replied:
oh OK makes sense, thanks!

jerryzh168 committed:
[quant][pt2e][be] Remove _input_output_share_observers and _reuse_input_obs_or_fq from QuantizationAnnotation (#102854)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/102854

att, after we support SharedQuantizationSpec we don't need these things anymore, this PR refactors the
uses of _input_output_share_observers to SharedQuantizationSpec

Test Plan:
```
buck2 test mode/opt caffe2/test:quantization_pt2e -- 'caffe2/test:quantization_pt2e'
```

Reviewed By: andrewor14

Differential Revision: D46301342

fbshipit-source-id: b40f4b1d5fb40c36f5964f8b99ad2e981cbc2da2
